### ISSUE TITLE
chore: Add better error and warning handling to `forc-pkg`

### ISSUE THREAD

**Issue by eureka-cpu** [FuelLabs/sway#4817]
related to #4709 & #928 

`forc-pkg` currently uses anyhow for error handling but for warnings we are just printing to std out as soon as they are found. We could take notes from `sway_core::error` and have a dedicated warning and error handler.

Use `thiserror` for the errors themselves as stated by #1201 & #3283

**Comment by eureka-cpu** [comment]
@gr00vytvniks this issue will take the place of the others mentioned in this issue as it is more accurate to the change being made.

**Comment by kayagokalp** [comment]
This is part of #1201 and maybe a duplicate of #3283

**Comment by eureka-cpu** [comment]
Based on `sway-core` error handling, the `forc-pkg` error handler may look something like:
```rust
use std::path::Path;
use thiserror::Error as ThisError;

use crate::manifest::ManifestFile;

#[derive(ThisError, Debug)]
enum ForcPkgError {
    #[error("could not find manifest at path: {0}")]
    MissingManifest(PathBuf), // Forc.toml
}

enum ForcPkgWarning {
    UnusedManifestKey { path: Path },
}

impl fmt::Display for ForcPkgWarning {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Self::UnusedManifestKey { path } => {
                write!(f, "  Warning: unused manifest key {path}")
            }
        }
    }
}

#[derive(Debug, Clone)]
pub struct ForcPkgResult<T> {
    pub value: T,
    pub warnings: Vec<ForcPkgWarning>,
    pub errors: Vec<ForcPkgError>,
}

#[cfg(test)]
mod tests {
    fn test_forc_pkg_res() {
        assert_eq!(pkg_res_helper().errors.0, ForcPkgError::MissingManifest);
    }
}

fn pkg_res_helper() -> ForcPkgResult<ManifestFile> {
    ManifestFile::from_dir(Default::default())
}

```

**Comment by kayagokalp** [comment]
> @kayagokalp here's my thoughts if you wouldn't mind chiming in? **Disclaimer:** this is draft code and may not run
> 
> Based on `sway-core` error handling, the `forc-pkg` error handler might look something like:
> 
> ```rust
> use std::path::Path;
> use thiserror::Error as ThisError;
> 
> use crate::manifest::ManifestFile;
> 
> #[derive(ThisError, Debug)]
> enum ForcPkgError {
>     #[error("could not find manifest at path: {0}")]
>     MissingManifest(PathBuf), // Forc.toml
> }
> 
> enum ForcPkgWarning {
>     UnusedManifestKey { path: Path },
> }
> 
> impl fmt::Display for ForcPkgWarning {
>     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
>         match self {
>             Self::UnusedManifestKey { path } => {
>                 write!(f, "  Warning: unused manifest key {path}")
>             }
>         }
>     }
> }
> 
> #[derive(Debug, Clone)]
> pub struct ForcPkgResult<T> {
>     pub value: T,
>     pub warnings: Vec<ForcPkgWarning>,
>     pub errors: Vec<ForcPkgError>,
> }
> 
> #[cfg(test)]
> mod tests {
>     fn test_forc_pkg_res() {
>         assert_eq!(pkg_res_helper().errors.0, ForcPkgError::MissingManifest);
>     }
> }
> 
> fn pkg_res_helper() -> ForcPkgResult<ManifestFile> {
>     ManifestFile::from_dir(Default::default())
> }
> ```

@eureka-cpu Yeah this seems nice, feels like it is on a correct path. Are there any other warnings other than unused manifest key though? Why not return `Result<T, ForcPkgError>` from functions directly?

**Comment by eureka-cpu** [comment]
> @eureka-cpu Yeah this seems nice, feels like it is on a correct path. Are there any other warnings other than unused manifest key though? Why not return `Result<T, ForcPkgError>` from functions directly?

There are other warnings that `cargo` emits wrt the `Cargo.toml` that we don't for `Forc.toml` and possibly other warnings that may only be relevant to fuel packages. One that comes to mind is "manifests that are not part of a workspace but believe they are". I could be wrong here, totally okay with it either way!
